### DISCUSSION TITLE
dot autocomplete but only variables

### DISCUSSION THREAD

**Discussion by snowzky** [pydata/xarray#5137] [Q&A]
Hi, this is not a bug or really an issue, so it's super cool you guys have this more low-level Q&A-category, which I hope it's the right place for this question.

I use xarray mainly for loading in netcdf files, if its just timeline data of different variables I sent it to Pandas, and if it's map data with a time axis, or just with no time, I remain in xarray (but don't know yet how to all the stuff I want) - and other than that I use xarray for easy and quick exploration, which this question is about.

I am still very much a beginner, so bear that in mind.

**Is there a way to limit dot-notation tab-completion to only variables, not attributes and methods?**

I love using dot notation in my IDE, just typing a few letters and then tab-autocomplete
`ds.pre[TAB] --> ds.precipitation_estimate.plot.hist()`
to get a quick plot to learn more about my data. I used pandas before xarray, and really enjoy the same here.
Or just leave it at `ds.pre[TAB] --> ds.precipitation_estimate` to get that sneak-peak at one's data.

The particular reason I am asking is that for this main dataset I happen to be working on, *alot* of my attributes and also some methods happen to start with the same few letters as some variables. This just happens to be the case for the attributes, but the methods are always there.

I looked around and I thought maybe the ds.variables would be an "accessor" to tab-autocomplete for only variables, but no.
Or have I missed such a function?

So don't get me wrong, I love the autocomplete functionality, but there are *so* many things directly onto the dataset that . will tab-autocomplete, that it sometimes makes it less useful.

**Comment by max-sixty** [comment]
Thanks for the question @snowzky

At the moment `ds.variables` is a frozen dict, which operates like a dict and so doesn't have `.{key}` access.

I think we could implement [`_ipython_key_completions_`](https://ipython.readthedocs.io/en/stable/config/integrating.html ) for the frozen dict, and get completions on `ds.variables["<TAB>`.

If that's correct, we would definitely take a PR for that.

**Reply by keewis** [reply]
`ds["<TAB>` already works with IPython and autocompletes variable names, file names and cell / line magic (not sure if it is possible to disable everything except variable names)
